lib/std/sort.zig

```diff
@@ -89,23 +93,30 @@ pub fn heapContext(a: usize, b: usize, context: anytype) void {
     }
 }
 
-fn siftDown(a: usize, root: usize, n: usize, context: anytype) void {
-    var node = root;
-    while (true) {
-        var child = a + 2 * (node - a) + 1;
-        if (child >= n) break;
+fn siftDown(a: usize, cur: usize, b: usize, context: anytype) void {
+    var child = (math.mul(usize, cur - a, 2) catch return) + a + 1;
+    // When we don't overflow from the multiply, the above expression equals (2*cur) - (2*a) + a + 1
+    // The `+ a + 1` is safe because:
+    //  for `a > 0` then `2a >= a + 1`.
+    //  for `a = 0`, the expression equals `2*cur+1`. `2*cur` is an even number, therefore adding 1 is safe.
 
-        // choose the greater child.
-        child += @intFromBool(child + 1 < n and context.lessThan(child, child + 1));
+    // stop if we overshot the boundary
+    if (child < b) {} else return;
 
-        // stop if the invariant holds at `node`.
-        if (!context.lessThan(node, child)) break;
+    const next_child = child + 1; // `next_child` is at most `b`, therefore no overflow is possible
 
-        // swap `node` with the greater child,
-        // move one step down, and continue sifting.
-        context.swap(node, child);
-        node = child;
+    // store the greater child in `child`
+    if (next_child < b and context.lessThan(child, next_child)) {
+        child = next_child;
     }
+
+    // stop if the invariant holds at `cur`.
+    if (context.lessThan(child, cur)) return;
+
+    // swap `cur` with the greater child,
+    // move one step down, and continue sifting.
+    context.swap(child, cur);
+    return @call(.always_tail, siftDown, .{ a, child, b, context });
```

Comment:
The CI tells us that LLVM does not support tail calls on all targets. Is it possible to write the new logic as a loop instead?

Reply:
I changed it to a tail call because I thought it was cleaner because this version of the function does not need to modify a variable from the parameters/arguments. I will change it back.